File: q21_car_rental/solution-1-java/CarRentalService.java
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.*;

/**
 * In-memory Car Rental System with "early return frees remaining days" rule.
 *
 * Availability semantics:
 * - If a trip ends EARLY (actualEndDate < tillDate), the car is occupied through actualEndDate
 *   (inclusive) and becomes available from the NEXT day.
 * - If a trip ends LATE (actualEndDate > tillDate), the car remains occupied through actualEndDate
 *   (inclusive) — new bookings overlapping that extension must be rejected.
 *
 * Cost still uses: effectiveEndDate = max(tillDate, endDate).
 */
public class CarRentalService {

    // --- Entities ---

    private static final class Car {
        final String licensePlate;
        final int costPerDay;
        final int freeKmsPerDay;
        final int costPerKm;

        Car(String licensePlate, int costPerDay, int freeKmsPerDay, int costPerKm) {
            this.licensePlate = licensePlate;
            this.costPerDay = costPerDay;
            this.freeKmsPerDay = freeKmsPerDay;
            this.costPerKm = costPerKm;
        }
    }

    private static final class Booking {
        final String orderId;
        final String carPlate;
        final LocalDate fromDate;
        final LocalDate tillDate;

        Integer startOdometer;      // set at startTrip
        LocalDate actualEndDate;    // set at endTrip (input endDate)
        Integer finalOdometer;      // set at endTrip
        Integer totalCost;          // cached after first endTrip computation (idempotent)

        Booking(String orderId, String carPlate, LocalDate fromDate, LocalDate tillDate) {
            this.orderId = orderId;
            this.carPlate = carPlate;
            this.fromDate = fromDate;
            this.tillDate = tillDate;
        }
    }

    // --- Storage ---

    private final Map<String, Car> cars = new HashMap<>();
    private final Map<String, Booking> bookingsByOrder = new HashMap<>();
    private final Map<String, List<Booking>> bookingsByCar = new HashMap<>();

    public CarRentalService() { }

    // --- API Methods ---

    public void addCar(String licensePlate, int costPerDay, int freeKmsPerDay, int costPerKm) {
        if (licensePlate == null || licensePlate.isBlank()) return;
        if (costPerDay < 0 || freeKmsPerDay < 0 || costPerKm < 0) return;
        cars.putIfAbsent(licensePlate, new Car(licensePlate, costPerDay, freeKmsPerDay, costPerKm));
    }

    public boolean bookCar(String orderId, String carLicensePlate, String fromDate, String tillDate) {
        if (orderId == null || orderId.isBlank()) return false;
        if (carLicensePlate == null || carLicensePlate.isBlank()) return false;
        if (!cars.containsKey(carLicensePlate)) return false;
        if (bookingsByOrder.containsKey(orderId)) return false;

        LocalDate from = LocalDate.parse(fromDate);
        LocalDate till = LocalDate.parse(tillDate);
        if (till.isBefore(from)) return false; // fromDate ≤ tillDate

        // Overlap check (inclusive) with existing bookings for this car.
        // We consider the occupied range as [b.fromDate .. occupiedEndDate(b)].
        List<Booking> existing = bookingsByCar.getOrDefault(carLicensePlate, Collections.emptyList());
        for (Booking b : existing) {
            LocalDate occupiedEnd = occupiedEndDate(b);
            if (overlapsInclusive(from, till, b.fromDate, occupiedEnd)) {
                return false;
            }
        }

        Booking booking = new Booking(orderId, carLicensePlate, from, till);
        bookingsByOrder.put(orderId, booking);
        bookingsByCar.computeIfAbsent(carLicensePlate, k -> new ArrayList<>()).add(booking);
        return true;
    }

    public void startTrip(String orderId, int odometerReading) {
        Booking b = bookingsByOrder.get(orderId);
        if (b == null) return;
        b.startOdometer = odometerReading;
    }

    public int endTrip(String orderId, int finalOdometerReading, String endDate) {
        Booking b = bookingsByOrder.get(orderId);
        if (b == null) return -1;

        if (b.totalCost != null) {
            return b.totalCost;
        }

        Car car = cars.get(b.carPlate);
        LocalDate providedEnd = LocalDate.parse(endDate);
        LocalDate effectiveEnd = providedEnd.isAfter(b.tillDate) ? providedEnd : b.tillDate;

        int days = (int) (ChronoUnit.DAYS.between(b.fromDate, effectiveEnd) + 1); // inclusive
        int startOdo = (b.startOdometer != null) ? b.startOdometer : finalOdometerReading;
        int tripKms = Math.max(0, finalOdometerReading - startOdo);

        int freeAllowance = days * car.freeKmsPerDay;
        int extraKms = Math.max(0, tripKms - freeAllowance);

        int total = days * car.costPerDay + extraKms * car.costPerKm;

        // Persist trip closure details
        b.actualEndDate = providedEnd;   // availability derives from this (occupied through this date inclusive)
        b.finalOdometer = finalOdometerReading;
        b.totalCost = total;

        return total;
    }

    // --- Helpers ---

    /**
     * The car is considered occupied through the ACTUAL end date inclusive if the trip has ended,
     * otherwise through the originally booked tillDate.
     * - Early end (actualEndDate < tillDate): occupiedEnd = actualEndDate (free from the next day)
     * - Late end  (actualEndDate > tillDate): occupiedEnd = actualEndDate (block new bookings inside extension)
     * - Not ended: occupiedEnd = tillDate
     */
    private static LocalDate occupiedEndDate(Booking b) {
        return (b.actualEndDate != null) ? b.actualEndDate : b.tillDate;
    }

    // Inclusive overlap: [aStart..aEnd] overlaps [bStart..bEnd] iff aStart ≤ bEnd and bStart ≤ aEnd
    private static boolean overlapsInclusive(LocalDate aStart, LocalDate aEnd, LocalDate bStart, LocalDate bEnd) {
        return !aEnd.isBefore(bStart) && !bEnd.isBefore(aStart);
    }
}

File: q21_car_rental/solution-2-python/CarRentalService.py
from dataclasses import dataclass
from datetime import date
from typing import Dict, List, Optional


@dataclass(frozen=True)
class Car:
    license_plate: str
    cost_per_day: int
    free_kms_per_day: int
    cost_per_km: int


@dataclass
class Booking:
    order_id: str
    car_plate: str
    from_date: date
    till_date: date

    # Filled later
    start_odometer: Optional[int] = None
    actual_end_date: Optional[date] = None
    final_odometer: Optional[int] = None
    total_cost: Optional[int] = None


class CarRentalService:
    """
    Availability semantics for booking overlap checks:
    - Before endTrip: occupied [from_date .. till_date] (inclusive).
    - Early end (actual_end_date < till_date): occupied through actual_end_date only
      (car becomes available starting the NEXT day).
    - Late end  (actual_end_date > till_date): occupied through actual_end_date.
    Trip cost uses: effective_end_date = max(till_date, provided end_date at endTrip).
    """

    def __init__(self):
        self._cars: Dict[str, Car] = {}
        self._bookings_by_order: Dict[str, Booking] = {}
        self._bookings_by_car: Dict[str, List[Booking]] = {}

    # --- Utilities ---

    @staticmethod
    def _to_int(x) -> Optional[int]:
        try:
            return int(x)
        except (TypeError, ValueError):
            return None

    # --- API ---

    def addCar(self, licensePlate: str, costPerDay, freeKmsPerDay, costPerKm) -> None:
        if not licensePlate or licensePlate.strip() == "":
            return
        cpd = self._to_int(costPerDay)
        fkp = self._to_int(freeKmsPerDay)
        cpk = self._to_int(costPerKm)
        if cpd is None or fkp is None or cpk is None:
            return
        if cpd < 0 or fkp < 0 or cpk < 0:
            return
        if licensePlate in self._cars:
            return
        self._cars[licensePlate] = Car(licensePlate, cpd, fkp, cpk)

    def bookCar(self, orderId: str, carLicensePlate: str, fromDate: str, tillDate: str) -> bool:
        if not orderId or orderId.strip() == "":
            return False
        if not carLicensePlate or carLicensePlate.strip() == "":
            return False
        if carLicensePlate not in self._cars:
            return False
        if orderId in self._bookings_by_order:
            return False

        try:
            from_d = date.fromisoformat(fromDate)
            till_d = date.fromisoformat(tillDate)
        except ValueError:
            return False
        if till_d < from_d:
            return False  # fromDate ≤ tillDate

        # Overlap check (inclusive) against existing bookings for this car
        for b in self._bookings_by_car.get(carLicensePlate, []):
            occ_end = self._occupied_end_date(b)
            if self._overlaps_inclusive(from_d, till_d, b.from_date, occ_end):
                return False

        booking = Booking(orderId, carLicensePlate, from_d, till_d)
        self._bookings_by_order[orderId] = booking
        self._bookings_by_car.setdefault(carLicensePlate, []).append(booking)
        return True

    def startTrip(self, orderId: str, odometerReading) -> None:
        b = self._bookings_by_order.get(orderId)
        if b is None:
            return
        odo = self._to_int(odometerReading)
        if odo is None:
            return
        b.start_odometer = odo

    def endTrip(self, orderId: str, finalOdometerReading, endDate: str) -> int:
        b = self._bookings_by_order.get(orderId)
        if b is None:
            return -1
        if b.total_cost is not None:
            return b.total_cost  # idempotent

        odo_final = self._to_int(finalOdometerReading)
        if odo_final is None:
            return -1
        try:
            provided_end = date.fromisoformat(endDate)
        except ValueError:
            return -1

        car = self._cars[b.car_plate]

        # Charge up to the later of the two dates
        effective_end = provided_end if provided_end > b.till_date else b.till_date

        # Inclusive day count
        days = (effective_end - b.from_date).days + 1

        start_odo = b.start_odometer if b.start_odometer is not None else odo_final
        trip_kms = max(0, odo_final - start_odo)

        free_allowance = days * car.free_kms_per_day
        extra_kms = max(0, trip_kms - free_allowance)

        total = days * car.cost_per_day + extra_kms * car.cost_per_km

        # Persist trip closure details for future availability checks
        b.actual_end_date = provided_end
        b.final_odometer = odo_final
        b.total_cost = total

        return total

    # --- Helpers ---

    def _occupied_end_date(self, b: Booking) -> date:
        """
        Occupied end date is:
        - actual_end_date if the trip has ended (early or late),
        - otherwise till_date.
        This enables early-release from the day AFTER actual_end_date,
        and blocks late returns through the extended date.
        """
        return b.actual_end_date if b.actual_end_date is not None else b.till_date

    @staticmethod
    def _overlaps_inclusive(a_start: date, a_end: date, b_start: date, b_end: date) -> bool:
        # [a_start..a_end] overlaps [b_start..b_end] iff a_start ≤ b_end and b_start ≤ a_end
        return not (a_end < b_start or b_end < a_start)
